presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ParametricScalarImplementation.java

```diff
@@ -86,89 +93,69 @@
         implements ParametricImplementation
 {
     private final Signature signature;
-    private final boolean nullable;
-    private final List<ArgumentProperty> argumentProperties;
-    private final MethodHandle methodHandle;
-    private final List<ImplementationDependency> dependencies;
-    private final Optional<MethodHandle> constructor;
-    private final List<ImplementationDependency> constructorDependencies;
-    private final List<Class<?>> argumentNativeContainerTypes;
+    private final List<Optional<Class<?>>> argumentNativeContainerTypes; // argument native container type is Optional.empty() for function type
     private final Map<String, Class<?>> specializedTypeParameters;
+    private final Class<?> returnNativeContainerType;
+    private final List<ParametricScalarImplementationChoice> choices;
 
-    public ParametricScalarImplementation(
+    private ParametricScalarImplementation(
             Signature signature,
-            boolean nullable,
-            List<ArgumentProperty> argumentProperties,
-            MethodHandle methodHandle,
-            List<ImplementationDependency> dependencies,
-            Optional<MethodHandle> constructor,
-            List<ImplementationDependency> constructorDependencies,
-            List<Class<?>> argumentNativeContainerTypes,
-            Map<String, Class<?>> specializedTypeParameters)
+            List<Optional<Class<?>>> argumentNativeContainerTypes,
+            Map<String, Class<?>> specializedTypeParameters,
+            List<ParametricScalarImplementationChoice> choices,
+            Class<?> returnContainerType)
     {
         this.signature = requireNonNull(signature, "signature is null");
-        this.nullable = nullable;
-        this.argumentProperties = ImmutableList.copyOf(requireNonNull(argumentProperties, "argumentProperties is null"));
-        this.methodHandle = requireNonNull(methodHandle, "methodHandle is null");
-        this.dependencies = ImmutableList.copyOf(requireNonNull(dependencies, "dependencies is null"));
-        this.constructor = requireNonNull(constructor, "constructor is null");
-        this.constructorDependencies = ImmutableList.copyOf(requireNonNull(constructorDependencies, "constructorDependencies is null"));
         this.argumentNativeContainerTypes = ImmutableList.copyOf(requireNonNull(argumentNativeContainerTypes, "argumentNativeContainerTypes is null"));
         this.specializedTypeParameters = ImmutableMap.copyOf(requireNonNull(specializedTypeParameters, "specializedTypeParameters is null"));
+        this.choices = requireNonNull(choices, "choices is null");
+        this.returnNativeContainerType = requireNonNull(returnContainerType, "return native container type is null");
+
+        for (int i = 1; i < choices.size(); i++) {
+            checkCondition(Objects.equals(choices.get(i).getDependencies(), choices.get(0).getDependencies()), FUNCTION_IMPLEMENTATION_ERROR, "Implementation dependencies do not match");
```

Comment:
`Implementations for the same function signature must have matching dependencies: <signature>`

Ditto for the next two.